lib/qmp.ml

```diff
@@ -99,126 +99,156 @@ module Event = struct
   let _XEN_PLATFORM_PV_DRIVER_INFO = "XEN_PLATFORM_PV_DRIVER_INFO"
 end
 
+(* class associated with a error message from an Invalid JSON syntax *)
+let _JSONParsing = "JSONParsing"
 
-let message_of_string x =
-  let int = function
-  | `Int x -> x
-  | _ -> failwith "int" in
-  (* let float = function
-  | `Int x -> float_of_int x
-  | _ -> failwith "float" in *)
-  let string = function
-  | `String x -> x
-  | _ -> failwith "string" in
-  let assoc = function
-  | `Assoc x -> x
-  | _ -> failwith "assoc" in
-  let bool = function
-  | `Bool x -> x
-  | _ -> failwith "bool" in
-  match Yojson.Safe.from_string x with
-  | `Assoc
-     [ ("QMP", `Assoc [ ("version", `Assoc [ "qemu", `Assoc version; "package", `String package ]); ("capabilities", _)] )] ->
-    Greeting {
-      minor = int (List.assoc "minor" version);
-      major = int (List.assoc "major" version);
-      micro = int (List.assoc "micro" version);
-      package = package;
+let message_of_string str =
+  let module Y = Yojson.Safe in
+  let module U = Yojson.Safe.Util in
+
+  let for_all xs args = List.for_all (fun x->List.mem x args) xs in
```

Comment:
Predicate `for_all xs args` is true, if all elements in `xs` are found in `args`. This means, `xs` is a subset of `args`. A better name would be `contained_in` or `subset_of`. The current name does not reflect that `for_all xs ys != for_all ys xs`.